README.md
# mwanova - Multi-Way Analysis of Variance.

**mwanova** is a simple program to compute analysis of variance. It is distributed under the GNU General Public License. You should read the file COPYING before installing and running mwanova.

Before compiling take a look at *conf.h* and redefine any variables you need. Most of them control the sizes of static variables for data and computed values. Be careful if you need to change them.

This version of **mwanova** is a revision of the former mwanova-1.1, including an almost complete rewrite of the code which abolished some of its previous limitations (and introduced some bugs). **mwanova** has still some limitations as any other piece of software. First of all, it  can only deal with as much as 10 factors. I can hardly imagine an experiment with 10 or more factors, but if you need one, just change the definitions MAXFACTORS in the *conf.h* file (hope you have enough RAM for that...).

The maximum number of data values (*MAXDATA*), levels per factor (*MAXLEVELS*) and combinations (*COMBINS*) have been eliminated. *COMBINS* is still used but it does not limit the analysis up to 10 factors. Previously, *COMBINS* was the size of an array of all possible factor combinations for a 10 factor analysis, and was set to 1024. An analysis with 11 factors would have 2048 combinations and was only
possible if this constant was set to 2048. The current version of **mwanova**  creates a dynamically linked list of terms and the only limitation is RAM. However, *COMBINS* is still used to compute some string sizes to accommodate variance estimates of terms used in Cornfield-Tukey Rules. If you want to test a model with more than 10 *random* factors you should consider reading *model.cpp* and changing the value of *COMBINS* in  *conf.h*. Note that this is only true for an analysis with RANDOM factors. An analysis of 20 FIXED factors is easily accomplished by mwanova (well, it takes a while) without any change of *COMBINS* (but you must change *MAXFACTORS*).

**mwanova** can now handle automatically missing data! However, missing level combinations are not allowed...

**mwanova** does not depend anymore on external libraries for computation  of probabilities. This was possible thanks to Daniel A. Atkinson who developed CCMATH, an excellent library from where portions of code were grabbed. These reside in "probs.cpp". Igor Baskir also contributed with an algorithm to compute Cochran's C probabilities.

## Installation

Installation of mwanova is straightforward. First unpack the source using

`$ tar zxf mwanova-version.tar.gz`
	
This will make a directory called mwanova-<version> with the source code inside. Just cd into mwanova-<version>

`$ cd mwanova-version`
   
To compile and install mwanova type

`> autoreconf --install`

`> ./configure`

`> make`

`> make install`

Optionally you can configure mwanova to compile as a CGI executable with 

`> ./configure --enable-gci`

and the manual in /usr/local/man/man1 typing

`> make install-man`

(you should do this as root).


This is the second public release of mwanova, and since I have several minor versions (I started it in 1996), I have named it version 2.0. This version is a beta release. I've fixed all the bugs I was able to detect, but probably there are many more. If you find a bug, don't hesitate, and e-mail me. Suggestions are also welcome. Note, however, that I don't guarantee a reply. Currently, I have not much time to do bug fixes. The code is available, you know...

I've only tested **mwanova** in a Linux system, because I couldn't find a C++ compiler in our HP-UX... if you port the program to other architectures please tell me.

The algorithm is a bit complex. I'm no professional programmer or statistician. I don't have a clue of matrix algebra, but probably most things should have been done using matrixes. Anyway, mwanova works, and it works fine for me. I've compared its results with the results from some commercial packages and everything was OK (I've only tested fully orthogonal and fully nested models, but as mixed models can be derived from orthogonal models, I did also some tests for mixed models and they were all OK). Now mwanova must pass through all other possible combinations of data and models, and this is only possible if several different people use it.

In the directory data, an example of a data file and model file will produce the following results:

`> mwanova -f Data/data.dat`

```
------------------------ ANOVA RESULTS ---------------------------------
Source of Variation        SS  DF          MS           F       P  Against 
------------------------------------------------------------------------
F1                     46.482   1      46.482       7.345   0.019  Residual
F2                      0.082   1       0.082       0.013   0.911  Residual
F1xF2                  11.207   1      11.207       1.771   0.208  Residual
F3                      4.148   2       2.074       0.328   0.727  Residual
F1xF3                   8.651   2       4.325       0.684   0.523  Residual
F2xF3                   0.766   2       0.383       0.061   0.942  Residual
F1xF2xF3                1.486   2       0.743       0.117   0.890  Residual
Residual               75.940  12       6.328
------------------------------------------------------------------------
TOTAL                 148.760  23   
```         

If you find the program useful, please e-mail me telling so. Don't forget to cite it if you use mwanova in any published paper... thanks, and enjoy it. 

## DONE TO DO's

- [x] Handling of missing data automatically!

- [x] Multiple comparison tests (Tukey, SNK, Schaffe) included in the program!

- [x] Better ANOVA table with terms ordered by importance (main factors, first order interactions, etc).
    
- [x] Incorporated probability functions in the program!
        
## STILL TO DO (if I can find some time to...)

- [ ] Handling of "Non-Testable" F-Ratios automatically

- [ ] Compute asymmetrical designs (not unbalanced data-sets!). Several experiments demand crossed factors and a single external control, not crossed or nested in the other factors. I must understand this first, because I only know how to compute this for a one-way ANOVA (but it is extensible to more factors).

- [ ] Compute BACI (Before-After, Control versus Impact) designs... Very useful when dealing with impact assessment data. Guess this is not easy...

## License
Copyright (C) 2001  Antonio Santos (amsantos@fc.up.pt)

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA




ChangeLog
2003-11-07  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.am (SUBDIRS): Add m4.
	(ACLOCAL_AMFLAGS): New variable.
	(EXTRA_DIST): New variable.
	* configure.in (AC_OUTPUT): Add po/Makefile.in,

31-7-2003 Version 2.0.2

Another anoying and hidden bug fixed. This prevented good designs with
two orthogonal factors nested into another one to be computed. The problem
was only revealed when the nested factor with the highest number of levels 
was entered in a column before the other nested factor and not the other
way...

7-6-2001 Version 2.0.1

A "hidden" bug was found! This affected seriously the final results
but only if the data included a nested factor with more than 10 levels
and observations were entered in an unordered way. Ordered datasets
were immune to the bug. The  problem was in the algorithm used to
orthogonalize the data. A new algorithm, with recursive functions
(which I was avoiding) seems to be working properly with any type of
datasets.

Another minor bug was found in the functions which compute the
Cornfield-Tukey rules. In some very particular cases (complex mixed models) some terms had no tests when, in fact, there were tests for them. 

24-5-2001 Version 2.0

SNK tests added. Homogeneity tests now include exact probabilities
thanks to Igor Baskir who has provided a way to compute Cochran's C
probabilities. F and Chi-square probabilities do not depend anymore
on external libraries. The code to compute such probabilities was
grabbed from CCMATH, developed by Daniel A. Atkinson to whom I also
want to thank.

29-3-2001 Version 2.0-pre1

Complete rewrite of code. Most things are now handled dynamically.
Multiple comparison tests included. Handling of missing data included.
Input files now contain all information about the model... no more
model files.

19-6-1999 Version 1.1

Fixed a bug in means.cpp which prevented mwanova from computing
means tables for more than 5 factors!

Fixed a small bug in CT Rules related with mixed models. For some mixed
models, where a factor was nested in another but orthogonal to a third
one, mwanova produced a correct CT Table but displayed non-testable
F ratios when there was, in fact, a test.

Fixed a small problem in Makefile related with the absence of libprob

20-1-1999 Version 1.0

Option -a now produces a table that also includes variance estimates
for each average.

Option -a now produces some Homosdeascity tests but no probabilities, because
there are no Hartley's Fmax, Cochran's C, etc probs in the net.


src/probs.cpp
// probs.cpp
//
// mwanova - Multi-Way Analysis of Variance.
// Copyright (C) 2001  Antonio Santos (amsantos@fc.up.pt)
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// *************************************************************************
// 
// This file contains code to compute probabilities of several statistics
// It is mainly a collection of code of the excellent lib CCMATH 2.2.0
// developed by Daniel A. Atkinson (DanAtk@aol.com). Applied Statistics
// algorithms were used for the computation of studentized range
// probabilities. Igor Baskir (baskir@univer.kharkov.ua) is the author
// of the code that computes Cochran's C probabilities.

#include <cmath>
#include "conf.h"
using namespace std;

//------------------------------------------------------------------------//
// Algorithm AS66 Applied Statistics (1973) vol22 no.3                    //
// Evaluates the tail area of the standardised normal curve               //
// from x to infinity if upper is .true. or                               //
// from minus infinity to x if upper is .false.                           //
//------------------------------------------------------------------------//
#include "probs.h"
using namespace std;

double alnorm(double x)
{
 const double ltone  = 7.0;
 const double utzero = 18.66;
 const double con    = 1.28;
 const double p      = 0.398942280444;
 const double q      = 0.39990348504;
 const double r      = 0.398942280385;
 const double a1     = 5.75885480458;
 const double a2     = 2.62433121679;
 const double a3     = 5.92885724438;
 const double b1     =-29.8213557807;
 const double b2     = 48.6959930692;
 const double c1     =-3.8052E-8;
 const double c2     = 3.98064794E-4;
 const double c3     =-0.151679116635;
 const double c4     = 4.8385912808;
 const double c5     = 0.742380924027;
 const double c6     = 3.99019417011;
 const double d1     = 1.00000615302;
 const double d2     = 1.98615381364;
 const double d3     = 5.29330324926;
 const double d4     =-15.1508972451;
 const double d5     = 30.789933034;

 double z,y,res;
 bool up;

 up = true;
 z  = x;
 if(!(z >= 0)){
   up = !(up);
   z = -z;
 }
 if((z <= ltone) || up && (z <=utzero)){
   y = 0.5*z*z;
   if(z > con)  res = r*exp(-y)/(z+c1+d1/(z+c2+d2/(z+c3+d3/(z+c4+d4/(z+c5+d5/(z+c6))))));
   else res = 0.5-z*(p-q*y/(y+a1+b1/(y+a2+b2/(y+a3))));
 } 
 else res = 0;
 if(!(up)) res = 1-res;
 return res;
}


//------------------------------------------------------------------------//
// Algorithm AS 190  Appl. Statist. (1983) Vol.32, No. 2                  //
// Incorporates corrections from Appl. Statist. (1985) Vol.34 (1)         //
// Evaluates the probability from 0 to q for a studentized                //  
// range having v degrees of freedom and r samples.                       // 
// Uses subroutine ALNORM = algorithm AS66.                               //
//                                                                        //
// Arrays vw and qw store transient values used in the                    //
// quadrature summation.  Node spacing is controlled by                   //
// step.  pcutj and pcutk control truncation.                             //
// Minimum and maximum number of steps are controlled by                  //
// jmin, jmax, kmin and kmax.  Accuracy can be increased                  //
// by use of a finer grid - Increase sizes of arrays vw                   // 
// and qw, and jmin, jmax, kmin, kmax and 1/step proportionally.          //
// of freedom. Uses AS prtrng and alnorm                                  //
//------------------------------------------------------------------------//

double prtrng(double q, int v, int r) 
{
 short  int ifault;
 double retprtrng;
 double vw[31], qw[31];
 double vmax=120.0;
 double cv1=0.193064705;
 double cv2=0.293525326;
 double cvmax=0.39894228;
 double cv[5]={0, 0.318309886, -0.268132716e-2, 0.347222222e-2, 0.833333333e-1};
 double g, gmid, r1, c, h, v2, gstep, pk1, pk2, gk, pk;
 double w0, pz, x, hj, ehj, pj;
 double pcutj=0.00003, pcutk=0.0001, step=0.45;
 int  jmin=3, jmax=15, kmin=7, kmax=15;
 int  k, j, jj, jump;

  

  /* check initial values */
  retprtrng=0.0;
  ifault=0;
  if ((v < 1.0) || (r < 2.0)) ifault=1;
  if ((q <= 0.0) || (ifault == 1)) return retprtrng;

  /* computing constants, local midpoint, adjusting steps */
  g=step * pow(r, -0.2);
  gmid=0.5 * log((double)r);
  r1=r-1.0;
  c=log(r*g*cvmax);
  if (v <= vmax){ 
  
   h=step * pow(v, -0.5);
   v2=v * 0.5;
   if (v == 1.0) c=cv1;
   if (v == 2.0) c=cv2;
   if (!((v == 1.0) || (v == 2.0)) ) c=sqrt(v2) * cv[1] /(1.0 + ((cv[2]/v2 + cv[3])/ v2 + cv[4]) / v2);
   c=log(c*r*g*h);  
  }
  
  /* computing integral */
at20:
  gstep=g;
  qw[1]=(-1.0);
  qw[jmax+1]=(-1.0);
  pk1=1.0;
  pk2=1.0;

  for (k=1; k<=kmax; k++) {
    gstep=gstep-g;
  at21:
    gstep=(-1.0)*gstep;
    gk=gmid + gstep;
    pk=0.0;
    if ((pk2 <= pcutk) && (k > kmin)) goto at26;
    w0=c-gk*gk*0.5;
    pz=alnorm(gk);
    x=alnorm(gk-q) - pz;
    if (x > 0.0) pk=exp(w0+r1*log(x));
    if (v > vmax) goto at26;

    jump=(-1)*jmax;
  at22:
    jump=jump + jmax;
    for (j=1; j<=jmax; j++) {
      jj=j+jump;
      if (qw[jj] > 0.0) goto at23;
      hj=h*j;
      if (j < jmax) qw[jj+1]=(-1.0);
      ehj=exp(hj);
      qw[jj]=q*ehj;
      vw[jj]=v*(hj+0.5-ehj*ehj*0.5);
    at23:
      pj=0.0;
      x=alnorm(gk-qw[jj]) - pz;
      if (x > 0.0) pj=exp(w0+vw[jj]+r1*log(x));
      pk=pk+pj;
      if (pj > pcutj) goto at24;
      if ((jj > jmin) || (k> kmin)) goto at25;
    at24:
      if (1==2) x=0; // dummy statement;
    }
  at25:
    h=(-1.0)*h;
    if (h < 0.0) goto at22;
at26:
    retprtrng=retprtrng + pk;
    if ((k > kmin) && (pk <= pcutk) && (pk1 <= pcutk)) goto ende;
    pk2=pk1;
    pk1=pk;
    if (gstep > 0.0) goto at21;
  }
  
ende:
  return(retprtrng);;
}

//------------------------------------------------------------------------//
//  Extracted from gaml.c    CCMATH mathematics library source code.      //
//                                                                        //
//  Copyright (C)  2000   Daniel A. Atkinson    All rights reserved.      //
//  This code may be redistributed under the terms of the GNU library     //
//  public license (LGPL). ( See the lgpl.license file for details.)      //
//------------------------------------------------------------------------//

double gaml(double x)
{ 
 double g,h;
 for(g=1.; x<30. ;g*=x,x+=1.); 
 h=x*x;
 g=(x-.5)*log(x)-x+.918938533204672-log(g);
 g+=(1.-(1./6.-(1./3.-1./(4.*h))/(7.*h))/(5.*h))/(12.*x);
 return g;
}

//------------------------------------------------------------------------//
//  Extracted from qgama.c    CCMATH mathematics library source code.     //
//                                                                        //
//  Copyright (C)  2000   Daniel A. Atkinson    All rights reserved.      //
//  This code may be redistributed under the terms of the GNU library     //
//  public license (LGPL). ( See the lgpl.license file for details.)      //
//------------------------------------------------------------------------//


double qgama(double x,double a)
{ 
 double ap,ro,f,t,gaml(double); int k;
 ro=a*log(x)-x-gaml(a);
 ap=6.25; if(a>ap) ap=a; t=(x-ap)/(f=sqrt(2.*ap));
 if(x<4.5 || t< -1. || (t< -.5 && a<10.)){
  for(f=t=1.,ap=a; t>1.e-14 ;){ t*=x/(ap+=1.); f+=t;}
  return 1.-exp(ro)*f/a;
 }
 else{
  if(t<0. && a<10.) k=18;
  else{ 
   if(t>3.){ k=(int) ceil(19.-3.*t); if(k<4) k=4;}
   else k=6+(int)ceil(f*(2.05-.8*t+.091*t*t));
  } 
  for(f=x; k>0 ;){ t=k--; f=x+(t-a)/(1.+t/f);}
  return exp(ro)/f;
 }
}


//------------------------------------------------------------------------//
//  Extracted from qbeta.c    CCMATH mathematics library source code.     //
//                                                                        //
//  Copyright (C)  2000   Daniel A. Atkinson    All rights reserved.      //
//  This code may be redistributed under the terms of the GNU library     //
//  public license (LGPL). ( See the lgpl.license file for details.)      //
//------------------------------------------------------------------------//

double qbeta(double x, double a, double b)
{ 
 double ro,t,ts,f,gaml(double); int nf;
 ro=gaml(a)+gaml(b)-gaml(a+b);
 if(x<.5) nf=1; else{ x=1.-x; t=a; a=b; b=t; nf=0;}
 f=t=exp(a*log(x)+b*log(1.-x)-ro)/a;
 for(ts=0.,b+=a-1.; t>1.e-12 || t>ts ;){
   b+=1.; a+=1.; ts=t; t*=x*b/a; f+=t; 
 }
 if(nf) return f; else return 1.-f;
}


//*********************************************************************//
//                           MAIN FUNCTIONS                            //
//*********************************************************************//

//------------------------------------------------------------------------//
// This function computes F exact probability for v1 and v2  degrees of   //
// freedom. Uses CCMATH's qbeta.                                          //
//------------------------------------------------------------------------//

double fprob(double f, int n1, int n2)
{
 return qbeta(n2/(n2+n1*f),0.5*n2,0.5*n1);
}

//------------------------------------------------------------------------//
// This function computes the inverse of the F distribution for v1 and v2 //
//  degrees of freedom. Uses CCMATH's qbeta.                              //
//------------------------------------------------------------------------//

double fiprob(double f, int n1, int n2)
{
 double z;
 z=qbeta(0.95,(double)0.5*n2,(double) 0.5*n1);
 return (n2*(1-z))/(n1*z);
}

//------------------------------------------------------------------------//
// This function computes Q exact probability for k means and df degrees  // 
// of freedom. Uses AS prtrng and alnorm.                                 //
//------------------------------------------------------------------------//

double qprob(double q, int k, int df)
{
 return 1-prtrng(q,df,k);
}

//------------------------------------------------------------------------//
// This function computes Cochran's C exact probability for k means and   //
// df degrees of freedom.                                                 //
// I must thank Igor Baskir <baskir@univer.kharkov.ua> for this algorithm.//
//------------------------------------------------------------------------//

double cprob(double c, int k, int df)
{
 double P=0;
 if((c>0)&&(k>1)){
  P=fprob((1.0/c-1.0)/(double)(k-1.0),((k-1)*df),df)*k;
  if(P>1) P=fabs(ceil(P)-P);
 }
 return P;
}


//------------------------------------------------------------------------//
// This function computes Chi-square exact probability of X^2 with        // 
// df degrees of freedom.                                                 //
//------------------------------------------------------------------------//


double chiprob(double c, int df)
{ 
 if((c<0.0)||(df<1.0)) return 0;
 return qgama(c*0.5,(double) df*0.5);
}




